framework/infrastructure/gcp/cloud_run.py

```diff
@@ -0,0 +1,147 @@
+import logging
+from typing import Any, Dict, Final
+
+from googleapiclient import discovery
+
+from framework.infrastructure import gcp
+
+logger = logging.getLogger(__name__)
+
+# Type aliases
+GcpResource = gcp.compute.ComputeV1.GcpResource
+
+DEFAULT_TEST_PORT: Final[int] = 8080
+DISCOVERY_URI: Final[str] = "https://run.googleapis.com/$discovery/rest?"
+
+
+@dataclasses.dataclass(frozen=True)
+class CloudRun:
+    service_name: str
+    url: str
+
+    @classmethod
+    def from_response(cls, name: str, d: Dict[str, Any]) -> "CloudRun":
+        return cls(
+            service_name=name,
+            url=d["urls"],
+        )
+
+
+class CloudRunApiManager(
+    gcp.api.GcpStandardCloudApiResource, metaclass=abc.ABCMeta
+):
+    project: str
+    region: str
+    _parent: str
+    service: discovery.Resource
+    api_manager: gcp.api.GcpApiManager
+
+    def __init__(self, project: str, region: str):
+        if not project:
+            raise ValueError("Project ID cannot be empty or None.")
+        if not region:
+            raise ValueError("Region cannot be empty or None.")
+        self.api_manager = gcp.api.GcpApiManager(v2_discovery_uri=DISCOVERY_URI)
+        self.project = project
+        self.region = region
+        service: discovery.Resource = self.api_manager.cloudrun("v2")
+        self.service = service
+        self._parent = f"projects/{self.project}/locations/{self.region}"
+        super().__init__(self.service, project)
+
+    @property
+    def api_name(self) -> str:
+        """Returns the API name for Cloud Run."""
+        return "run"
+
+    @property
+    def api_version(self) -> str:
+        """Returns the API version for Cloud Run."""
+        return "v2"
+
+    def create_service(
+        self, service: discovery.Resource, service_name: str, body: dict
+    ) -> GcpResource:
+        return self._create_resource(
+            collection=service.projects().locations().services(),
+            location=self.region,
+            serviceId=service_name,
+            body=body,
+        )
+
+    def get_service(
+        self, service: discovery.Resource, service_name: str
+    ) -> CloudRun:
+        result = self._get_resource(
+            collection=service.projects().locations().services(),
+            full_name=self.resource_full_name(
+                service_name, "services", self.region
+            ),
+        )
+        return CloudRun.from_response(
+            self.resource_full_name(service_name, "services", self.region),
+            result,
+        )
+
+    def get_service_uri(self, service_name: str) -> str:
+        response = self.get_service(self.service, service_name)
+        return response.url
+
+    def delete_service(self, service_name: str):
+        # pylint: disable=no-member
+        self._delete_resource(
+            self.service.projects().locations().services(),
+            full_name=self.resource_full_name(
+                service_name, "services", self.region
```

Comment:
make "services" a class constant, like in https://github.com/grpc/psm-interop/blob/12922820b1d662b8a1b7a11d9199c4c606a49923/framework/infrastructure/gcp/network_security.py#L160